l1-contracts/contracts/bridge/AssetRouterBase.sol

```diff
@@ -0,0 +1,268 @@
+        emit BridgehubDepositBaseTokenInitiated(_chainId, _prevMsgSender, _assetId, _amount);
+    }
+
+    /// @notice Initiates a deposit transaction within Bridgehub, used by `requestL2TransactionTwoBridges`.
+    /// @param _chainId The chain ID of the ZK chain to which deposit.
+    /// @param _prevMsgSender The `msg.sender` address from the external call that initiated current one.
+    /// @param _value The `msg.value` on the target chain tx.
+    /// @param _data The calldata for the second bridge deposit.
+    /// @return request The data used by the bridgehub to create L2 transaction request to specific ZK chain.
+    function bridgehubDeposit(
+        uint256 _chainId,
+        address _prevMsgSender,
+        uint256 _value,
+        bytes calldata _data
+    )
+        external
+        payable
+        virtual
+        override
+        onlyBridgehub
+        whenNotPaused
+        returns (L2TransactionRequestTwoBridgesInner memory request)
+    {
+        (bytes32 assetId, bytes memory transferData) = abi.decode(_data[1:], (bytes32, bytes));
+
+        require(BRIDGE_HUB.baseTokenAssetId(_chainId) != assetId, "AR: baseToken deposit not supported");
+
+        bytes memory bridgeMintCalldata = _burn({
+            _chainId: _chainId,
+            _value: _value,
+            _assetId: assetId,
+            _prevMsgSender: _prevMsgSender,
+            _transferData: transferData
+        });
+        bytes32 txDataHash = keccak256(bytes.concat(bytes1(0x01), abi.encode(_prevMsgSender, assetId, transferData)));
+
+        request = _requestToBridge({
+            _chainId: _chainId,
+            _prevMsgSender: _prevMsgSender,
+            _assetId: assetId,
+            _bridgeMintCalldata: bridgeMintCalldata,
+            _txDataHash: txDataHash,
+            _deposit: true
+        });
+
+        emit BridgehubDepositInitiated({
+            chainId: _chainId,
+            txDataHash: txDataHash,
+            from: _prevMsgSender,
+            assetId: assetId,
+            bridgeMintCalldata: bridgeMintCalldata
+        });
+    }
+
+    /// @notice Finalize the withdrawal and release funds.
+    /// @param _chainId The chain ID of the transaction to check.
+    /// @param _assetId The bridged asset ID.
+    /// @param _transferData The position in the L2 logs Merkle tree of the l2Log that was sent with the message.
+    function finalizeTransfer(
+        uint256 _chainId,
+        bytes32 _assetId,
+        bytes calldata _transferData
+    ) public virtual override returns (address l1Receiver, uint256 amount) {
+        address assetHandler = assetHandlerAddress[_assetId];
+
+        if (assetHandler != address(0)) {
+            IAssetHandler(assetHandler).bridgeMint(_chainId, _assetId, _transferData);
+        } else {
+            IAssetHandler(address(nativeTokenVault)).bridgeMint(_chainId, _assetId, _transferData); // ToDo: Maybe it's better to receive amount and receiver here? transferData may have different encoding
```

Comment:
We're returning data from finalizeTransfer 
The main question is:
1. why're we doing?
2. we expect specific transferData decoding, however it might differ for custom AHs

If we decide to still return data, let's instead allow AH to return the values

Reply:
Lets just return the correct values in the legacy case and 0 values otherwise.